src/org/traccar/CoordinatesHandler.java

```diff
@@ -39,10 +39,12 @@ private Position getLastPosition(long deviceId) {
     @Override
     protected Position handlePosition(Position position) {
         Position last = getLastPosition(position.getDeviceId());
-        if (last != null) {
+        if (last != null && last.getValid() && last.getLatitude() != 0 && last.getLongitude() != 0) {
             double distance = DistanceCalculator.distance(
                     position.getLatitude(), position.getLongitude(), last.getLatitude(), last.getLongitude());
-            if (distance < coordinatesMinError || distance > coordinatesMaxError) {
+            boolean satisfiesMin = coordinatesMinError == 0 || distance > coordinatesMinError;
+            boolean satisfiesMax = coordinatesMaxError == 0 || distance < coordinatesMaxError || position.getValid();
+            if (!(satisfiesMin && satisfiesMax)) {
```

Comment:
Sorry for being so picky, but can we please expand this if expressions to be
```java
!satisfiesMin || !satisfiesMax
```
I think it would be more readable like this.